gusto/equations.py

```diff
@@ -1347,24 +1370,38 @@ def __init__(self, domain, parameters, Omega=None,
         # -------------------------------------------------------------------- #
         # Pressure Gradient Term
         # -------------------------------------------------------------------- #
-        pressure_gradient_form = subject(prognostic(-div(w)*p*dx, 'u'), self.X)
+        pressure_gradient_form = pressure_gradient(
+            subject(prognostic(-div(w)*p*dx, 'u'), self.X))
 
         # -------------------------------------------------------------------- #
         # Gravitational Term
         # -------------------------------------------------------------------- #
-        gravity_form = subject(prognostic(-b*inner(w, domain.k)*dx, 'u'), self.X)
+        gravity_form = name_label(subject(prognostic(-b*inner(w, domain.k)*dx,
+                                                     'u'), self.X), "gravity")
 
         # -------------------------------------------------------------------- #
         # Divergence Term
         # -------------------------------------------------------------------- #
-        # This enforces that div(u) = 0
-        # The p features here so that the div(u) evaluated in the "forcing" step
-        # replaces the whole pressure field, rather than merely providing an
-        # increment to it.
-        divergence_form = name_label(
-            subject(prognostic(phi*(p-div(u))*dx, 'p'), self.X),
-            "incompressibility"
-        )
+
+        if compressible:
+            cs = parameters.cs
+            linear_div_form = name_label(
+                subject(prognostic(cs**2 * phi * div(u_trial) * dx, 'p'),
+                        self.X), "divergence")
+            divergence_form = name_label(linearisation(
+                subject(prognostic(cs**2 * phi * div(u) * dx, 'p'), self.X),
+                linear_div_form), "divergence")
+        else:
+            # This enforces that div(u) = 0
```

Comment:
the results look odd when running incompressible Boussinesq with RK4 but are fine with SIQN - this comment is about the forcing term which makes me suspicious...

Reply:
I'd assume it should work fine for small enough `dt` - the issue I was having was that the range of stable `dt` for the IMEX RK schemes was being limited by effectively treating a `div(u)` term explicitly. You could try using IMEX RK, with the div term labeled first explicitly, then implicitly and see if you see the same issues when it is explicit (but not when it is implicit). Or just use smaller `dt`?

Reply:
no, it's not that - sorry, I should have been clearer. The results don't show instability - they're just wrong. And the reason is that this isn't really an evolution equation for the pressure. What we're trying to do is solve for the pressure that gives us a non-divergent velocity. When we use the SIQN time stepper, the incompressible term is implicit and, as the comment here says, we overwrite the pressure rather than updating it.